cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go

```diff
@@ -1393,23 +1419,76 @@ func (p *BicepProvider) loadParameters(ctx context.Context) (map[string]azure.Ar
+	ParametersJson string `json:"parametersJson"`
+}
+
+type compileBicepResult struct {
+	RawArmTemplate azure.RawArmTemplate
+	Template       azure.ArmTemplate
+	Parameters     azure.ArmParameters
+}
+
 func (p *BicepProvider) compileBicep(
 	ctx context.Context, modulePath string,
-) (azure.RawArmTemplate, azure.ArmTemplate, error) {
-	compiled, err := p.bicepCli.Build(ctx, modulePath)
+) (*compileBicepResult, error) {
+	var err error
+	paramsType, err := p.parameterFile(ctx)
 	if err != nil {
-		return nil, azure.ArmTemplate{}, fmt.Errorf("failed to compile bicep template: %w", err)
+		return nil, err
+	}
+
+	var compiled string
+	var parameters azure.ArmParameters
+	if paramsType.fileType == bicepparam {
+		azdEnv := p.env.Environ()
+		// append principalID (not stored to .env by default). For non-bicepparam, principalId is resolved
+		// without looking at .env
+		if _, exists := p.env.LookupEnv(environment.PrincipalIdEnvVarName); !exists {
+			currentPrincipalId, err := p.curPrincipal.CurrentPrincipalId(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("fetching current principal id for bicepparam compilation: %w", err)
+			}
+			azdEnv = append(azdEnv, fmt.Sprintf("%s=%s", environment.PrincipalIdEnvVarName, currentPrincipalId))
+		}
+		compiled, err = p.bicepCli.BuildBicepParam(ctx, paramsType.filePath, azdEnv)
+		if err != nil {
+			return nil, fmt.Errorf("failed to compile bicepparam template: %w", err)
+		}
+
+		var bicepParamOutput compiledBicepParamResult
+		if err := json.Unmarshal([]byte(compiled), &bicepParamOutput); err != nil {
+			log.Printf("failed unmarshalling compiled bicepparam (err: %v), template contents:\n%s", err, compiled)
+			return nil, fmt.Errorf("failed unmarshalling arm template from json: %w", err)
+		}
+		compiled = bicepParamOutput.TemplateJson
+		var params azure.ArmParameterFile
+		if err := json.Unmarshal([]byte(bicepParamOutput.ParametersJson), &params); err != nil {
+			log.Printf("failed unmarshalling compiled bicepparam parameters(err: %v), template contents:\n%s", err, compiled)
+			return nil, fmt.Errorf("failed unmarshalling arm parameters template from json: %w", err)
+		}
+		parameters = params.Parameters
+	} else {
+		compiled, err = p.bicepCli.Build(ctx, modulePath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to compile bicep template: %w", err)
+		}
 	}
 
 	rawTemplate := azure.RawArmTemplate(compiled)
 
 	var template azure.ArmTemplate
 	if err := json.Unmarshal(rawTemplate, &template); err != nil {
 		log.Printf("failed unmarshalling compiled arm template to JSON (err: %v), template contents:\n%s", err, compiled)
-		return nil, azure.ArmTemplate{}, fmt.Errorf("failed unmarshalling arm template from json: %w", err)
+		return nil, fmt.Errorf("failed unmarshalling arm template from json: %w", err)
 	}
 
-	return rawTemplate, template, nil
+	return &compileBicepResult{
+		RawArmTemplate: rawTemplate,
+		Template:       template,
+		Parameters:     parameters,
```

Comment:
The `Parameters` field is used later by the `bicepProvider` to know if the `compileBicepResult` is coming from a `bicepparam` (because the field is already populated and ready to be used), or if it is coming from `.bicep` file (because the field is nil and azd still needs to parse the parameters.json and do its magic).